### PR TITLE
[SD-1296] - CMS Pages

### PR DESCRIPTION
**Pull request by MatthewKennedy**
This PR allows admins to create **Standard Pages**, **Feature Pages** and **Homepages** on a per-store basis with multiple language support.

### Standard Pages 
Standard pages are simple text pages that contain text entered using a Rich Text Editor in the admin UI, a typical example of a standard page would be an About Us, or Privacy Policy page.

### Feature Pages
Feature pages allow store owners to create pages containing multiple sections, such as hero images with buttons, promo sections with linking images,  a taxon of products displayed as a carousel of images,  featured articles, a featured product, and many more. Sections can be rearranged via drag and drop in the admin UI.

A real world example of a feature page in action would be when browsing Apple products on their website, before you get to picking your new iMac colour and specs, you are presented with a feature page that highlights unique selling points and hammers home the need to purchase using large splash images and short, to the point info bytes.  Then when you click the Buy button, you are then taken from the feature page to the product page to select your specs and add to bag.

### Homepages
Homepages are much like feature pages in the sense that you have many rearrangeable sections of various types, allowing you to build your homepage. The key difference is that admins are not allowed to give homepages a custom slug.

## To Do: 
- [x] Merge Menu Improvements into this PR when ready and re-attach linkable items from the new setup, make sure they work on menu items and sections.
- [x] Setup TinyMCE as gem.
- [x] Move sections to use STI for validation, image requirements and linking requirements on a per section basis.
- [x] Hook up lazy loading images for various image sizes on a per-section basis.
- [x] Hook up ability to link to pages from menu items.
- [x] Setup pages API
- [x] Add remaining sections from stock home page and create new ones.
- [x] Check for missing translations in front and backend UI.
- [x] add removable class to menu-items for fade-out when deleted.
- [x] Fix fresh_when for pages and home
- [x] Re-evaluate Admin CSS for mobile use on page section editor. (get the best possible use case on small devices).
- [x] Move some of the logic from the views to view helpers both front and backend.
- [x] Clean up code.
- [x] Document image sizes and aspect ratios for each section that uses images.
- [x] Re-evaluate database indexes and add where needed.
- [x] Write tests tests tests.
- [x] Write sample data for Homepages all demo stores,  add some sample standard pages with filler text.
- [x] Check `rel="canonical"` tags on homepage & cms_pages
- [x] Fix CC issues where possible
- [x] Fix handling of locale switching on cms_pages.
- [x] Admin UI should only offer store supported_locales for language options.
- [x] Add CSS for various screen sizes to change the zoom level of the Live Preview for large screens.
- [x] Add for API tests **reposition** and **draft_mode** actions.
- [x] Add filters on cms_pages admin index view. filter by locale and page type.
- [x] check sample data for menus is adding promotions as link not containers.
- [x] Fix linkable for pages to use slug - see `concerns/spree/display_link.rb` and add `.slug` for pages
- [x] Re-factor sections to be more use cases than those originally set on homepage.
- [x] Add ability to link to more than taxon for sections with multiple link points.
- [x] Tidy up locals and organise for easy use
- [x] set up tests on each section type when finalised.
- [x] Kill access to sections that require menus until Spree has ability to add menu locations.
- [x] Fix Rails 5.2 issue with _changed? on Active::Record::Store
- [ ] Add feature tests for Frontend and Backend for each type of Section.
- [x] Add request tests for Storefront API
- [ ] Update Storefront API OpenAPI docs
- [X] Fix code review issues.

## Known Issues To Fix After Per Store Admin Implementation 
- Page preview buttons need to check that they redirect to the correct domain name for the page and store they belong to.
- Live Preview test that on a per store basis.

### CODE REVIEW

File: core/db/migrate/20210512191732_create_spree_cms_pages.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class CreateSpreeCmsPages < ActiveRecord::Migration[6.0]
+  def change
+    create_table :spree_cms_pages do |t|
+      t.column :title, :string, null: false
+      t.column :meta_title, :string
+      t.column :content, :text
+      t.column :meta_description, :text
+      t.column :visible, :boolean, default: false
+      t.column :slug, :string
```
Comment (damianlegawiec) [inline core/db/migrate/20210512191732_create_spree_cms_pages.rb:9]: @MatthewKennedy let's create on index on `slug` and another index on `slug` + `store_id` + `locale`

File: frontend/app/controllers/spree/cms_pages_controller.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    include Spree::FrontendHelper
+    include Spree::CacheHelper
+
+    def show
+      if @page&.viewable?
+        @cms_page = @page
+      elsif try_spree_current_user&.admin?
+        @cms_page = @page
+        @edit_mode = true
+      end
```
Comment (damianlegawiec) [inline frontend/app/controllers/spree/cms_pages_controller.rb:14]:
```suggestion
      else
         raise ActiveRecord::RecordNotFound
      end
```

File: core/db/migrate/20210512191732_create_spree_cms_pages.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
+class CreateSpreeCmsPages < ActiveRecord::Migration[6.0]
+  def change
+    create_table :spree_cms_pages do |t|
+      t.column :title, :string, null: false
+      t.column :meta_title, :string
+      t.column :content, :text
+      t.column :meta_description, :text
+      t.column :visible, :boolean, default: false
```
Comment (damianlegawiec) [inline core/db/migrate/20210512191732_create_spree_cms_pages.rb:8]: @MatthewKennedy let's add an index for this

Comment (MatthewKennedy) [reply core/db/migrate/20210512191732_create_spree_cms_pages.rb:8]: @damianlegawiec Not a problem, I'll be on with this over the weekend.

File: api/app/controllers/spree/api/v2/storefront/cms_pages_controller.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
+module Spree
+  module Api
+    module V2
+      module Storefront
+        class CmsPagesController < ::Spree::Api::V2::ResourceController
+          private
+
```
[the remark this answers is not in the document]
Comment (MatthewKennedy) [reply api/app/controllers/spree/api/v2/storefront/cms_pages_controller.rb:7]: @damianlegawiec I think that should be taken care of with the last couple of commits, the default scope for cos_pages in the api controller is scoped by store locale and additionally as linkable, this ensures the type is not homepage and it has a slug.

File: api/app/controllers/spree/api/v2/platform/cms_pages_controller.rb [hunk 5]
```diff
@@ -0,0 +1,33 @@
+module Spree
+  module Api
+    module V2
+      module Platform
+        class CmsPagesController < ResourceController
+          before_action -> { doorkeeper_authorize! :write, :admin }, only: WRITE_ACTIONS << :toggle_visibility
+
+          def toggle_visibility
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/platform/cms_pages_controller.rb:8]: @MatthewKennedy can we handle this by a standard update action?

File: api/app/controllers/spree/api/v2/storefront/cms_pages_controller.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+          def model_class
+            Spree::CmsPage
+          end
+
+          def resource
+            @resource ||= scope.find_by(slug: params[:slug])
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/cms_pages_controller.rb:13]:
```suggestion
            @resource ||= scope.friendly.find(params[:slug])
```

this way we can support both slugs and IDs

File: api/app/controllers/spree/api/v2/storefront/cms_pages_controller.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+          def resource_serializer
+            Spree::Api::Dependencies.storefront_cms_page_serializer.constantize
+          end
+
+          def collection_serializer
+            Spree::Api::Dependencies.storefront_cms_page_serializer.constantize
+          end
+
+          def scope
+            super.by_store(current_store).by_locale(I18n.locale).linkable
```
Comment (damianlegawiec) [inline api/app/controllers/spree/api/v2/storefront/cms_pages_controller.rb:25]:
```suggestion
            super.by_locale(I18n.locale).linkable
```

store scope is already applied

File: api/app/serializers/spree/v2/storefront/cms_section_serializer.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    module Storefront
+      class CmsSectionSerializer < BaseSerializer
+        set_type :cms_section
+
+        attributes :name, :content, :settings, :link, :fit, :type, :position
+
+        attribute :image_one_path do |section|
+          if section.image_one.attached?
+            url_helpers = Rails.application.routes.url_helpers
+            url_helpers.rails_blob_path(section.image_one, only_path: true)
```
Comment (damianlegawiec) [inline api/app/serializers/spree/v2/storefront/cms_section_serializer.rb:12]: @MatthewKennedy why are we returning paths and not absolute URLs?

Comment (damianlegawiec) [reply api/app/serializers/spree/v2/storefront/cms_section_serializer.rb:12]: we should keep this in sync how we return image URLs for products etc

File: api/config/routes.rb [hunk 9]
```diff
@@ -166,6 +166,9 @@
 
         resources :menus, only: %i[index]
         get '/menus/:location', to: 'menus#show', as: :menu
+
+        resources :cms_pages, only: %i[index], path: '/pages'
+        get '/pages/:slug', to: 'cms_pages#show', as: :page
```
Comment (damianlegawiec) [inline api/config/routes.rb:171]:
```suggestion
        resources :cms_pages, only: %i[index]
        get '/cms_pages/:slug', to: 'cms_pages#show', as: :page
```

this way we're avoiding any name clashing

File: core/app/models/spree/cms_page.rb [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+    has_many :menu_items, as: :linked_resource
+
+    before_validation :handle_slug
+
+    validates :title, :store, :locale, presence: true
+    validates :slug, uniqueness: { scope: :store, allow_nil: true, case_sensitive: true }
+
+    scope :visible, -> { where(visible: true) }
+    scope :by_store, ->(store) { where(store: store) }
```
Comment (damianlegawiec) [inline core/app/models/spree/cms_page.rb:20]: @MatthewKennedy we already have `for_store` scope

File: core/app/models/spree/cms_page.rb [hunk 11]
```diff
@@ -0,0 +1,64 @@
+module Spree
+  class CmsPage < Spree::Base
+    include Spree::DisplayLink
```
Comment (damianlegawiec) [inline core/app/models/spree/cms_page.rb:3]:
```suggestion
    include SingleStoreResource
    include Spree::DisplayLink
```

File: core/app/models/spree/store.rb [hunk 12]
```diff
@@ -13,6 +13,7 @@ class Store < Spree::Base
     has_many :store_payment_methods, class_name: 'Spree::StorePaymentMethod'
     has_many :payment_methods, through: :store_payment_methods, class_name: 'Spree::PaymentMethod'
 
+    has_many :cms_pages, class_name: 'Spree::CMSPage'
```
Comment (damianlegawiec) [inline core/app/models/spree/store.rb:16]: @MatthewKennedy we also need to as cms_sections association

File: core/app/models/spree/store.rb [hunk 13]
```diff
@@ -96,6 +97,22 @@ def supported_currencies_list
       end.uniq.compact
     end
 
+    def homepage(requested_locale)
+      Spree::CmsPage.find_by(store_id: id, locale: requested_locale, type: 'Spree::Cms::Pages::Homepage') ||
```
Comment (damianlegawiec) [inline core/app/models/spree/store.rb:101]:
```suggestion
      cms_pages.by_locale(requested_locale).where(type: 'Spree::Cms::Pages::Homepage') ||
```

File: core/app/models/spree/store.rb (hunk 12 above)
Comment (damianlegawiec) [inline core/app/models/spree/store.rb:16]:
```suggestion
    has_many :cms_pages, class_name: 'Spree::CmsPage'
```

File: core/app/models/spree/cms_page.rb (hunk 11 above)
Comment (MatthewKennedy) [reply core/app/models/spree/cms_page.rb:4]: This is done, I don't know why it isn't  picking up as changed.

File: api/app/controllers/spree/api/v2/platform/cms_pages_controller.rb (hunk 5 above)
Comment (MatthewKennedy) [reply api/app/controllers/spree/api/v2/platform/cms_pages_controller.rb]: This now uses a simple update on the visible attribute.

**Pull request merged**